src/python/pants/util/objects.py

```diff
@@ -523,6 +523,9 @@ class TypedCollection(TypeConstraint):
   """A `TypeConstraint` which accepts a TypeOnlyConstraint and validates a collection."""
 
   _iterable_constraint = SubclassesOf(Iterable)
+  # strings in Python are considered iterables of substrings, but we only want to allow explicit
+  # collection types.
+  _exclude_iterable_constraint = SubclassesOf(*six.string_types)
```

Comment:
Preferred to use `SubclassesOf(str, bytes)` with the relevant `builtins` imports. We want to avoid `six.string_types` as much as possible and to always make it explicit, as the point of Py3 is to minimize as much as possible the mixing of bytes and unicode.

Reply:
I remembered you had discussed why `six` is not a good library to do anything with, but hadn't recalled why. The justification here for being explicit makes sense.

Reply:
Six is especially bad because it has problematic backports, namely `StringIO`, which behaves differently on Py2 vs. Py3.

We could here use `from future.utils import string_types`, but we should still avoid it even though Future is great as explicit is much better than `string_types`.